### DISCUSSION TITLE
What is the best practice to have long living global effects?

### DISCUSSION THREAD

**Discussion by Paul-Svetlichny** [pointfreeco/swift-composable-architecture#1864] [Q&A] [answered]
We are now migrating our application into using TCA and I'm quite new to the subject, so my question might be answered already, but I can't find where, so will be thankful for pointing in the right direction.
In the app we have a global property that is used throughout the whole app. We use set of repositories (managers), which use CoreData and other frameworks to store and update this value. Combine is used to publish value to other view models and/or views. We use websockets (socket-io) to update this value from outside of the app, pushed by our BE.
So at some point early in the app we create some publishers, which are responsible for saving/retrieving/updating that value, and these publishers persist through the whole application lifecycle.

What is the proper way of handling this in TCA?
I've created a store with actions, which listen to the publisher and it works, but when I try to test it with TestStore, compiler argues that the effect is still running, which is obviously true as it never finishes.

I have a feeling that it might be an architectural mistake on our end or if it should be handled differently with TCA, but I have no idea in which direction to look. So I will appreciate any advise, thank you!
Of course, I can provide any additional information if what I described is not enough to give an advise.

Have a great day!

**Comment by mbrandonw** [comment, marked as answer]
Hi @Paul-Svetlichny, I don't think you are using the library wrong or misunderstanding  how things should be structured. While it is true that TCA's testing tools like you to exhaustively prove everything about the feature, including that all effects finish, sometimes that isn't possible. If you truly have an effect that lives for the entire duration of the app, then there is no user action that causes the effect to stop (other than force-quitting the app).

So, you have two choices:

* You can actually get a handle on the task that represents the lifetime of the effect in the test, and then cancel it once your test is complete:
  ```swift
  let task = store.send(.appDidLaunch)
  …
  await task.cancel() // App shuts down
  ```
  That should quiet the test failure.

* Alternatively you could try out some of our ["non-exhaustive" testing tools](https://www.pointfree.co/blog/posts/83-non-exhaustive-testing-in-the-composable-architecture), which allow you to just assert on the things you care about.

**Reply by Paul-Svetlichny** [reply]
Thanks for the hint @mbrandonw! It should be exactly what I was looking for!
